persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateQueryHandler.java

```diff
@@ -236,81 +251,275 @@ private void authorizeByToken(ByteBuffer token, CQLStatement statement) {
       }
 
       logger.debug(
-          "preparing to authorize statement of type {} on {}.{}",
+          "authorized statement of type {} on {}", castStatement.getClass().toString(), resource);
+    } else if (statement instanceof ListRolesStatement) {
+      ListRolesStatement stmt = (ListRolesStatement) castStatement;
+      String role = getRoleResourceFromStatement(stmt, "grantee");
+      logger.debug(
+          "preparing to authorize statement of type {} on {}",
           castStatement.getClass().toString(),
-          keyspaceName,
-          tableName);
+          role);
 
       try {
-        authorization.authorizeSchemaWrite(authToken, keyspaceName, tableName, scope);
+        authorization.authorizeRoleRead(authToken, role);
       } catch (io.stargate.auth.UnauthorizedException e) {
         throw new UnauthorizedException(
-            String.format(
-                "Missing correct permission on %s.%s",
-                keyspaceName, (tableName == null ? "" : tableName)));
+            String.format("Missing correct permission on role %s", role), e);
       }
 
       logger.debug(
-          "authorized statement of type {} on {}.{}",
+          "authorized statement of type {} on {}", castStatement.getClass().toString(), role);
+    } else if (statement instanceof ListPermissionsStatement) {
+      ListPermissionsStatement stmt = (ListPermissionsStatement) castStatement;
+      String role = getRoleResourceFromStatement(stmt, "grantee");
+      logger.debug(
+          "preparing to authorize statement of type {} on {}",
           castStatement.getClass().toString(),
-          keyspaceName,
-          tableName);
+          role);
+
+      try {
+        authorization.authorizePermissionRead(authToken, role);
+      } catch (io.stargate.auth.UnauthorizedException e) {
+        throw new UnauthorizedException(
+            String.format("Missing correct permission on role %s", role), e);
+      }
+
+      logger.debug(
+          "authorized statement of type {} on {}", castStatement.getClass().toString(), role);
+    }
+  }
+
+  private String getRoleResourceFromStatement(Object stmt, String fieldName) {
+    try {
+      Class<?> aClass = stmt.getClass();
+      if (stmt instanceof ListUsersStatement
+          || stmt instanceof GrantPermissionsStatement
+          || stmt instanceof RevokePermissionsStatement) {
+        aClass = aClass.getSuperclass();
+      }
+
+      Field f = aClass.getDeclaredField(fieldName);
+      f.setAccessible(true);
+      RoleResource roleResource = (RoleResource) f.get(stmt);
+
+      return roleResource != null ? roleResource.getName() : null;
+    } catch (Exception e) {
+      logger.error("Unable to get " + fieldName, e);
+      throw new RuntimeException("Unable to get private field", e);
+    }
+  }
+
+  private String getResourceFromStatement(PermissionsManagementStatement stmt) {
+    try {
+      Field f = stmt.getClass().getSuperclass().getDeclaredField("resource");
```

Comment:
the `resource` is protected. Maybe instead of accessing it using reflection, we can create a class that extracts it?
We could create an `org.apache.cassandra.cql3.statements;` package in our project and extract it from there?

Reply:
Seems like a lot of copy pasted code for a single field since we would have to copy pretty much every class that implements CQLStatement across C* 3.11, C* 4.0, and DSE 6.8. We can revisit later if this becomes and issue.